src/odm-collection.hpp

```diff
@@ -0,0 +1,656 @@
+* @param v A BSON document view. If the BSON document does not match the schema of type T,
+* BSONArchiver will throw a corresponding exception.
+* TODO what does BSONArchiver throw?
+* @return An object that corresponds to the given document view.
+*/
+template <class T>
+T to_obj(bsoncxx::document::view v) {
+    bson_mapper::bson_istream bis(v);
+    // TODO cereal::BSONInputArchive archive(bis);
+    in_archiver_mock<T> archive(bis);
+    T obj;
+    archive(obj);
+    return obj;
+}
+
+/*
+* This function converts an stdx::optional containing a BSON document value into an stdx::optional
+* containing a deserialized object of the templated type.
+* @tparam T A type that is serializable to BSON using a BSONArchiver.
+* @param opt An optional BSON document view
+* @return   An optional object of type T
+*/
+template <class T>
+mongocxx::stdx::optional<T> to_optional_obj(
+    const mongocxx::stdx::optional<bsoncxx::document::value>& opt) {
+    if (opt) {
+        T obj = to_obj<T>(opt.value().view());
+        return mongocxx::stdx::optional<T>(obj);
+    } else {
+        return mongocxx::stdx::optional<T>();
+    }
+}
+
+template <class T>
+class deserializing_cursor;
+
+template <class T>
+class odm_collection {
+   public:
+    odm_collection(mongocxx::collection& c) noexcept;
+
+    odm_collection(odm_collection&&) noexcept;
+    odm_collection& operator=(odm_collection&&) noexcept;
+    ~odm_collection();
+
+    /**
+     * Returns a copy of the underlying collection.
+     * @return The mongocxx::collection that this object wraps
+     */
+    mongocxx::collection collection();
+
+    ///
+    /// Runs an aggregation framework pipeline against this collection, and returns the results as
+    /// de-serialized object.
+    ///
+    /// @param pipeline
+    ///   The pipeline of aggregation operations to perform.
+    /// @param options
+    ///   Optional arguments, see mongocxx::mongocxx::options::aggregate.
+    ///
+    /// @return A deserializing_cursor with the results.
+    /// @throws
+    ///   If the operation failed, the returned cursor will throw an exception::query
+    ///   when it is iterated.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/command/aggregate/
+    ///
+    deserializing_cursor<T> aggregate(
+        const mongocxx::pipeline& pipeline,
+        const mongocxx::options::aggregate& options = mongocxx::options::aggregate());
+
+    ///
+    /// Counts the number of documents matching the provided filter.
+    ///
+    /// @param filter
+    ///   The filter, as a serializable object, that documents must match in order to be counted.
+    /// @param options
+    ///   Optional arguments, see mongocxx::options::count.
+    ///
+    /// @return The count of the documents that matched the filter.
+    /// @throws exception::query if the count operation fails.
+    ///
+    /// @see http://docs.mongodb.org/manual/reference/command/count/
+    ///
+    std::int64_t count(T filter,
```

Comment:
I'm not sure if we want `T filter` for now over `bsoncxx::document::view_or_value filter`. We should see how other ODM's handle this, but I think for now a view_or_value is fine.